cabal-install/Distribution/Client/IndexUtils.hs

```diff
@@ -139,6 +142,33 @@ indexBaseName repo = repoLocalDir repo </> fn
 -- Reading the source package index
 --
 
+-- 'IndexState' defined in "Distribution.Client.IndexUtils.Timestamp"
+-- to avoid import cycles
+
+-- | Effective 'IndexState' after applying 'filerCache'
+data IndexStateInfo = IndexStateInfo
+    { isiMaxTime  :: !Timestamp -- ^ Effective 'IndexStateTime' state
+    , isiHeadTime :: !Timestamp -- ^ 'Timestamp' equivalent to 'IndexStateHead'
+    }
+
+emptyStateInfo :: IndexStateInfo
+emptyStateInfo = IndexStateInfo nullTimestamp nullTimestamp
+
+-- | 'IndexState' specification to 'Cache'
+--
+-- Note: 'filerCache' is idempotent in the 'Cache' value
+filterCache :: IndexState -> Cache -> (Cache, IndexStateInfo)
+filterCache IndexStateHead cache = (cache, IndexStateInfo{..})
+  where
+    isiMaxTime  = cacheHeadTs cache
+    isiHeadTime = cacheHeadTs cache
+filterCache (IndexStateTime ts0) cache0 = (cache, IndexStateInfo{..})
+  where
+    cache = Cache { cacheEntries = ents, cacheHeadTs = isiMaxTime }
+    isiHeadTime = cacheHeadTs cache0
+    isiMaxTime  = maximum (nullTimestamp:map cacheEntryTimestamp ents)
```

Comment:
We've had this twice now iirc, could pull it out as `maximumTimestamp` in the `Timestamp` module.